compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -356,6 +357,96 @@ simplifyExpr world = fst . cata go
       -- e    ==>    e
       e -> (embed (fmap fst e), infoStep world (fmap snd e))
 
+-- Top-down inliner for simple functions.
+inlineExpr :: World -> Expr -> Expr
+inlineExpr _world = go
+  where
+    mkApp :: Expr -> Expr -> Expr
+    mkApp (ELocation l e1) e2 = ELocation l (ETmApp e1 e2)
+    mkApp e1 e2 = ETmApp e1 e2
+
+    -- Special cases. We split the general cases into a separate function
+    -- 'goExpr' so that GHC's pattern exhaustiveness checker works better.
+    go :: Expr -> Expr
+    go = \case
+        EApps (EVal x) [TyArg _, TyArg _, TyArg _, TmArg f, TmArg g, TmArg h]
+            | qualModule x == ModuleName ["GHC", "Base"]
+            , qualObject x == ExprValName "."
+            -> go (mkApp f (mkApp g h))
+
+        EApps (EVal x) [TyArg _, TyArg _, TyArg t, TmArg f, TmArg g]
+            | qualModule x == ModuleName ["GHC", "Base"]
+            , qualObject x == ExprValName "."
+            , let y = freshenExprVar (freeVarsInExpr f <> freeVarsInExpr g) (ExprVarName "y")
+            -> go (ETmLam (y,t) (mkApp f (mkApp g (EVar y))))
```

Comment:
I just realized that this changes semantics if `f` or `g` have side effects.

Reply:
So we need safety information for `f` and `g` if we're going to inline `.`

So we really do need the bottom up information, and to try to recreate this in a bottom up fashion if possible.

... 

Back to the drawing board.